### PR TITLE
Adds a new configuration proxy.config.http.allow_multi_range

### PR DESCRIPTION
**Pull request by zwoop**
This is needed to prevent potential abuse with well formed multi-
range requests.

### CODE REVIEW

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -4428,24 +4428,47 @@ HttpSM::do_range_setup_if_necessary()
     do_range_parse(field);
 
     if (t_state.range_setup == HttpTransact::RANGE_REQUESTED) {
+      bool do_transform = false;
+
       if (!t_state.range_in_cache) {
         Debug("http_range", "range can't be satisfied from cache, force origin request");
```
Comment (oknet) [inline proxy/http/HttpSM.cc]: `t_state.num_range_fields` is `-1` here.

File: proxy/http/HttpSM.cc [hunk 2]
```diff
@@ -4428,24 +4428,47 @@ HttpSM::do_range_setup_if_necessary()
...
+      if (t_state.num_range_fields > 1) {
+        if (0 == t_state.txn_conf->allow_multi_range) {
+          t_state.range_setup = HttpTransact::RANGE_NONE;                                 // No Range required (not allowed)
+          t_state.hdr_info.client_request.field_delete(MIME_FIELD_RANGE, MIME_LEN_RANGE); // ... and nuke the Range header too
+        } else if (1 == t_state.txn_conf->allow_multi_range) {
+          do_transform = true;
+        } else {
+          t_state.range_setup = HttpTransact::RANGE_NOT_SATISFIABLE;
+        }
+      } else {
```
Comment (oknet) [inline proxy/http/HttpSM.cc]: `t_state.num_range_fields` is `1` here.

Comment (zwoop) [reply proxy/http/HttpSM.cc]: Hmm, you saying I should reset num_range_fields to 0 ?

File: proxy/http/HttpSM.cc (hunk 1 above)
Comment (zwoop) [reply proxy/http/HttpSM.cc]: Not sure what you are saying here. How can it be -1, if we have

```
t_state.range_setup == HttpTransact::RANGE_REQUESTED
```

This gets set, as far as I can tell, in HttpSM::parse_range_and_compare, no ?

Comment (oknet) [reply proxy/http/HttpSM.cc]: Within HttpSM::parse_range_and_compare, if

- parse failed, goto Lfaild;
- the number of valid range is not greater than 0, goto Lfaild;

At the Lfaild, `range_in_cache = false` and `num_range_fields = -1`.

`t_state.range_setup = HttpTransact::RANGE_REQUESTED` and `t_state.num_range_fields = nr` only if `nr > 0`.

The valid value of `num_range_fields` should be -1 or any number great than 0.

**Comment by zwoop** [comment]
Cheery-picked to 7.1.3

**Pull request merged**